lib/Conversion/CGGIToJaxite/CGGIToJaxite.cpp

```diff
@@ -0,0 +1,306 @@
+      : OpConversionPattern<lwe::TrivialEncryptOp>(context, /*benefit=*/2) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      lwe::TrivialEncryptOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    FailureOr<Value> result = getContextualJaxiteParams(op.getOperation());
+    if (failed(result)) return result;
+    Value serverParams = result.value();
+    lwe::EncodeOp encodeOp = op.getInput().getDefiningOp<lwe::EncodeOp>();
+    if (!encodeOp) {
+      return op.emitError() << "Expected input to TrivialEncrypt to be the "
+                               "result of an EncodeOp, but it was "
+                            << op.getInput().getDefiningOp()->getName();
+    }
+    auto createConstantOp = rewriter.create<jaxite::ConstantOp>(
+        op.getLoc(), op.getOutput().getType(), encodeOp.getPlaintext(),
+        serverParams);
+    rewriter.replaceOp(op, createConstantOp);
+    return success();
+  }
+};
+
+struct ConvertCGGIToJaxiteEncodeOp : public OpConversionPattern<lwe::EncodeOp> {
+  ConvertCGGIToJaxiteEncodeOp(mlir::MLIRContext *context)
+      : OpConversionPattern<lwe::EncodeOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      lwe::EncodeOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+class CGGIToJaxite : public impl::CGGIToJaxiteBase<CGGIToJaxite> {
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    auto *op = getOperation();
+
+    CGGIToJaxiteTypeConverter typeConverter(context);
+    RewritePatternSet patterns(context);
+    ConversionTarget target(*context);
+    addStructuralConversionPatterns(typeConverter, patterns, target);
+
+    target.addLegalDialect<jaxite::JaxiteDialect>();
+    target.addIllegalDialect<cggi::CGGIDialect>();
+    target.addIllegalDialect<lwe::LWEDialect>();
+    // FuncOp is marked legal by the default structural conversion patterns
+    // helper, just based on type conversion. We need more, but because the
+    // addDynamicallyLegalOp is a set-based method, we can add this after
+    // calling addStructuralConversionPatterns and it will overwrite the
+    // legality condition set in that function.
+    target.addDynamicallyLegalOp<func::FuncOp>([&](func::FuncOp op) {
+      int num_inputs = op.getFunctionType().getNumInputs();
+      if (num_inputs < 2) {
+        return false;
+      }
+      bool hasServerKeyArg = op.getFunctionType().getNumInputs() > 0 &&
+                             mlir::isa<jaxite::ServerKeySetType>(
+                                 op.getFunctionType().getInput(num_inputs - 2));
+      bool hasParamsArg = op.getFunctionType().getNumInputs() > 0 &&
+                          mlir::isa<jaxite::ParamsType>(
+                              op.getFunctionType().getInput(num_inputs - 1));
```

Comment:
Here you assert that there's an ordering of the serverkey/params arg in the function types - while the `getContextualJaxiteParams` checks the whole list instead of assuming the position in the list. Maybe you should just utilize the logic in `getContextualJaxiteParams` function here (or use the positional info in that func too?)?

Reply:
The ordering in the func args doesnt seem necessary since the ordering only matters in jaxite calls. Changed it to check the whole list via "HasArg" method. Lmk if there is a way to reuse getContextualJaxiteParams for this too?

Reply:
that lgtm! I think hasarg is pretty simple :) you could have tested if getContextualJaxiteParams returned failure but hasarg is much nicer